core/src/main/java/io/grpc/internal/DelayedStream.java

```diff
@@ -86,182 +95,224 @@ public PendingMessage(InputStream message, boolean shouldBeCompressed) {
     }
   }
 
-  DelayedStream(ClientStreamListener listener) {
-    this.listener = listener;
+  @Override
+  public void start(ClientStreamListener listener) {
+    synchronized (this) {
+      // start may be called at most once.
+      checkState(this.listener == null, "already started");
+      this.listener = checkNotNull(listener, "listener");
+
+      // Check error first rather than success.
+      if (error != null) {
+        listener.closed(error, new Metadata());
+      }
+      // In the event that an error happened, realStream will be a noop stream.  We still call
+      // start stream in order to drain references to pending messages.
+      if (unstartedStream != null) {
+        startStream();
+      }
+    }
   }
 
-  /**
-   * Creates a stream on a presumably usable transport. Must not be called if {@link
-   * #cancelledPrematurely}, as there is no way to close {@code stream} without double-calling
-   * {@code listener}. Most callers of this method will need to acquire the intrinsic lock to check
-   * {@code cancelledPrematurely} and this method atomically.
-   */
-  void setStream(ClientStream stream) {
+  @VisibleForTesting
```

Comment:
It's not used in the tests. So it should be private?
Since it's always called under the lock, I would remove the `synchronized` and annotate this method with `@GuardedBy("this")`.